File: Leetcode 146-LRU cache/lru_cache.go
package main

import (
	"fmt"
	"strings"
)

type LRUCache struct {
	capacity   int
	cache      map[int]*LinkedListNode
	linkedList *LinkedList
}

func Constructor(capacity int) LRUCache {
	cache := make(map[int]*LinkedListNode)
	return LRUCache{
		capacity:   capacity,
		cache:      cache,
		linkedList: &LinkedList{},
	}
}

func (this *LRUCache) Get(key int) int {
	if _, found := this.cache[key]; !found {
		return -1
	}

	value := this.cache[key].value
	this.linkedList.removeNode(this.cache[key])
	this.linkedList.insertAtTail(key, value)
	this.cache[key] = this.linkedList.getTail()

	return value
}

func (this *LRUCache) Put(key int, value int) {
	if _, found := this.cache[key]; !found {
		if this.linkedList.size >= this.capacity {
			this.linkedList.insertAtTail(key, value)
			this.cache[key] = this.linkedList.getTail()
			delete(this.cache, this.linkedList.getHead().key)
			this.linkedList.removeHead()
		} else {
			this.linkedList.insertAtTail(key, value)
			this.cache[key] = this.linkedList.getTail()
		}
	} else {
		this.linkedList.removeNode(this.cache[key])
		this.linkedList.insertAtTail(key, value)
		this.cache[key] = this.linkedList.getTail()
	}

}

func (this *LRUCache) printLinkedList() string {
	current := this.linkedList.head
	elements := []string{}
	for current != nil {
		elements = append(elements, fmt.Sprintf("%d", current.value))
		current = current.next
	}
	return strings.Join(elements, " -> ")
}

type LinkedListNode struct {
	key   int
	value int
	next  *LinkedListNode
	prev  *LinkedListNode
}

type LinkedList struct {
	size int
	head *LinkedListNode
	tail *LinkedListNode
}

func (l *LinkedList) removeNode(node *LinkedListNode) {
	if node == nil {
		return
	}
	if node.prev != nil {
		node.prev.next = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	}
	if node == l.head {
		l.head = l.head.next
	}
	if node == l.tail {
		l.tail = l.tail.prev
	}

	l.size -= 1
}

func (l *LinkedList) insertAtTail(key, value int) {
	newNode := &LinkedListNode{key: key, value: value}
	if l.tail == nil {
		l.head = newNode
		l.tail = newNode
	} else {
		l.tail.next = newNode
		newNode.prev = l.tail
		l.tail = newNode
	}

	l.size += 1
}

func (l *LinkedList) removeHead() {
	l.removeNode(l.head)
}

func (l *LinkedList) getHead() *LinkedListNode {
	return l.head
}

func (l *LinkedList) getTail() *LinkedListNode {
	return l.tail
}

func main() {
	lruCache := Constructor(2)
	lruCache.Put(1, 1)
	fmt.Println("Put(1, 1)")
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	lruCache.Put(2, 2)
	fmt.Println("\nPut(2, 2)")
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	fmt.Println("\nGet(1)")
	fmt.Println("value: ", lruCache.Get(1))
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	lruCache.Put(3, 3)
	fmt.Println("\nPut(3, 3)")
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	fmt.Println("\nGet(2)")
	fmt.Println("value: ", lruCache.Get(2))
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	lruCache.Put(4, 4)
	fmt.Println("\nPut(4, 4)")
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	fmt.Println("\nGet(1)")
	fmt.Println("value: ", lruCache.Get(1))
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	fmt.Println("\nGet(3)")
	fmt.Println("value: ", lruCache.Get(3))
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())

	fmt.Println("\nGet(4)")
	fmt.Println("value: ", lruCache.Get(4))
	fmt.Println("cache: ", lruCache.cache)
	fmt.Println("doubly linked list", lruCache.printLinkedList())
}

/*
output:
Put(1, 1)
cache:  map[1:0xc00011a000]
doubly linked list 1

Put(2, 2)
cache:  map[1:0xc00011a000 2:0xc00011a040]
doubly linked list 1 -> 2

Get(1)
value:  1
cache:  map[1:0xc00011a0a0 2:0xc00011a040]
doubly linked list 2 -> 1

Put(3, 3)
cache:  map[1:0xc00011a0a0 3:0xc00011a100]
doubly linked list 1 -> 3

Get(2)
value:  -1
cache:  map[1:0xc00011a0a0 3:0xc00011a100]
doubly linked list 1 -> 3

Put(4, 4)
cache:  map[3:0xc00011a100 4:0xc00011a1a0]
doubly linked list 3 -> 4

Get(1)
value:  -1
cache:  map[3:0xc00011a100 4:0xc00011a1a0]
doubly linked list 3 -> 4

Get(3)
value:  3
cache:  map[3:0xc00011a240 4:0xc00011a1a0]
doubly linked list 4 -> 3

Get(4)
value:  4
cache:  map[3:0xc00011a240 4:0xc00011a2a0]
doubly linked list 3 -> 4
*/

File: Leetcode 146-LRU cache/lru_cache.py
class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = {}
        self.linkedList = LinkedList()  

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cache:
            return -1
        
        value = self.cache[key].value
        self.linkedList.removeNode(self.cache[key])
        self.linkedList.insertAtTail(key, value)
        self.cache[key] = self.linkedList.getTail()
        
        return value
        

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key not in self.cache:
            if self.linkedList.size >= self.capacity:
                self.linkedList.insertAtTail(key, value)
                self.cache[key] = self.linkedList.getTail()
                del self.cache[self.linkedList.getHead().key]
                self.linkedList.removeHead()
            else:
                self.linkedList.insertAtTail(key, value)
                self.cache[key] = self.linkedList.getTail()
        else:
            self.linkedList.removeNode(self.cache[key])
            self.linkedList.insertAtTail(key, value)
            self.cache[key] = self.linkedList.getTail()
    
    def printLinkedList(self):
        current = self.linkedList.head
        elements = []
        while current is not None:
            elements.append(str(current.value))
            current = current.next
        return " -> ".join(elements)
        
class LinkedListNode:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None

class LinkedList:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None
    
    def removeNode(self, node):
        if node is None:
            return
        if node.prev is not None:
            node.prev.next = node.next
        if node.next is not None:
            node.next.prev = node.prev
        if node == self.head:
            self.head = self.head.next
        if node == self.tail:
            self.tail = self.tail.prev
        
        self.size -= 1
    
    def insertAtTail(self, key, value):
        newNode = LinkedListNode(key, value)
        if self.tail is None:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        
        self.size += 1
    
    def removeHead(self):
        self.removeNode(self.head)
    
    def getTail(self):
        return self.tail
    
    def getHead(self):
        return self.head


if __name__ == "__main__":
    lruCache = LRUCache(2)
    lruCache.put(1, 1)
    print("put(1, 1)")
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    lruCache.put(2, 2)
    print("\nput(2, 2)")
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    print("\nget(1)")
    print("value: ", lruCache.get(1))
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    lruCache.put(3, 3)
    print("\nput(3, 3)")
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    print("\nget(2)")
    print("value: ", lruCache.get(2))
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    lruCache.put(4, 4)
    print("\nput(4, 4)")
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    print("\nget(1)")
    print("value: ", lruCache.get(1))
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    print("\nget(3)")
    print("value: ", lruCache.get(3))
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

    print("\nget(4)")
    print("value: ", lruCache.get(4))
    print("cache: ", lruCache.cache)
    print("doubly linked list: ", lruCache.printLinkedList())

"""
output:
put(1, 1)
('cache: ', {1: <__main__.LinkedListNode instance at 0x107789cf8>})
('doubly linked list: ', '1')

put(2, 2)
('cache: ', {1: <__main__.LinkedListNode instance at 0x107789cf8>, 2: <__main__.LinkedListNode instance at 0x107789d40>})
('doubly linked list: ', '1 -> 2')

get(1)
('value: ', 1)
('cache: ', {1: <__main__.LinkedListNode instance at 0x107789d88>, 2: <__main__.LinkedListNode instance at 0x107789d40>})
('doubly linked list: ', '2 -> 1')

put(3, 3)
('cache: ', {1: <__main__.LinkedListNode instance at 0x107789d88>, 3: <__main__.LinkedListNode instance at 0x107789cf8>})
('doubly linked list: ', '1 -> 3')

get(2)
('value: ', -1)
('cache: ', {1: <__main__.LinkedListNode instance at 0x107789d88>, 3: <__main__.LinkedListNode instance at 0x107789cf8>})
('doubly linked list: ', '1 -> 3')

put(4, 4)
('cache: ', {3: <__main__.LinkedListNode instance at 0x107789cf8>, 4: <__main__.LinkedListNode instance at 0x107789d40>})
('doubly linked list: ', '3 -> 4')

get(1)
('value: ', -1)
('cache: ', {3: <__main__.LinkedListNode instance at 0x107789cf8>, 4: <__main__.LinkedListNode instance at 0x107789d40>})
('doubly linked list: ', '3 -> 4')

get(3)
('value: ', 3)
('cache: ', {3: <__main__.LinkedListNode instance at 0x107789d88>, 4: <__main__.LinkedListNode instance at 0x107789d40>})
('doubly linked list: ', '4 -> 3')

get(4)
('value: ', 4)
('cache: ', {3: <__main__.LinkedListNode instance at 0x107789d88>, 4: <__main__.LinkedListNode instance at 0x107789cf8>})
('doubly linked list: ', '3 -> 4')

"""

